docs/docs/contributors/FSharp.md

```diff
@@ -0,0 +1,128 @@
+
+For example [SynExpr.For](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synexpr.html#For), the definition looks like:
+
+```fsharp
+type SynExpr =
+    ...
+    /// F# syntax: 'for i = ... to ... do ...'
+    | For of
+        forDebugPoint: DebugPointAtFor *
+        toDebugPoint: DebugPointAtInOrTo *
+        ident: Ident *
+        equalsRange: range option *
+        identBody: SynExpr *
+        direction: bool *
+        toBody: SynExpr *
+        doBody: SynExpr *
+        range: range
+```
+
+  However, in Fantomas we have a partial active pattern that we use to easily grab the information we need from the AST.
+```fsharp
+let (|For|_|) =
+    function
+    | SynExpr.For (_, _, ident, equalsRange, e1, isUp, e2, e3, _) -> Some(ident, equalsRange, e1, e2, e3, isUp)
+    | _ -> None  
+```
+
+Notice the underscores, we don't use the `DebugPoint` info and `range`, so we drop that information in the result of the partial active pattern.
+We then later use this pattern to work with the AST we need.
+Example usage in `CodePrinter.fs`:
+
+```fsharp
+match expr with
+| For (ident, equalsRange, e1, e2, e3, isUp) ->
+    atCurrentColumn (
+        !- "for "
+        +> genIdent ident
+        +> genEq SynExpr_For_Equals equalsRange
+        +> sepSpace
+        +> genExpr astContext e1
+        +> ifElse isUp (!- " to ") (!- " downto ")
+        +> genExpr astContext e2
+        +> !- " do"
+        +> indent
+        +> sepNln
+        +> genExpr astContext e3
+        +> unindent
+    )
+| _ -> ...
+```
+
+- [Custom operators](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/operator-overloading#creating-new-operators). In F# there are some special operators like [|>](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-operators.html#(|%3E)) and [>>](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-operators.html#(%3E%3E)).  
+
+Note that these are just function themselves as well. Instead of specifying all the arguments after the function name, (infix) operators let you specify an argument before the operator and after. 
+
+In F#, you are able to create your own operators as well. In Fantomas, the most notable are `!-` and `+>`. We will cover them later, but if you peek in `CodePrinter.fs`, they are heavily used there.
+
+- [Signature files](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/signature-files).
+
+In Fantomas we use signature files to define the module boundaries. Everything that is both defined in the implementation file (the `*.fs` file) and in the signature file (the `*.fsi` file) is considered to be visible to other modules.
+If a signature file is present, there is no need to specify `private` in a function you don't want to be visible to other modules. Just don't add `val` entry to the signature file and it will be private automatically.
```

Comment:
```suggestion
If a signature file is present, there is no need to specify `private` in a function you don't want to be visible to other modules. Just don't add a `val` entry to the signature file and it will be private automatically.
```